File: slp/backend-dotnet/Features/Admin/AdminLogRepository.cs
using backend_dotnet.Data;
using Microsoft.EntityFrameworkCore;

namespace backend_dotnet.Features.Admin;

public class AdminLogRepository : IAdminLogRepository
{
    private readonly AppDbContext _db;

    public AdminLogRepository(AppDbContext db)
    {
        _db = db;
    }

    public async Task LogAsync(AdminLog log)
    {
        _db.AdminLogs.Add(log);
        await _db.SaveChangesAsync();
    }

    public async Task<(IEnumerable<AdminLog> Items, int TotalCount)> GetRecentAsync(AdminLogFilterDto filter, int page, int pageSize)
    {
        var query = _db.AdminLogs
            .Include(l => l.Admin)
            .AsQueryable();

        if (filter.From.HasValue)
            query = query.Where(l => l.CreatedAt >= filter.From.Value);
        if (filter.To.HasValue)
            query = query.Where(l => l.CreatedAt <= filter.To.Value);
        if (filter.AdminId.HasValue)
            query = query.Where(l => l.AdminId == filter.AdminId.Value);
        if (!string.IsNullOrWhiteSpace(filter.Action))
            query = query.Where(l => l.Action == filter.Action);
        if (!string.IsNullOrWhiteSpace(filter.TargetType))
            query = query.Where(l => l.TargetType == filter.TargetType);

        // Apply search in memory after fetching
        var logs = await query.OrderByDescending(l => l.CreatedAt).ToListAsync();

        if (!string.IsNullOrWhiteSpace(filter.Search))
        {
            var searchTerm = filter.Search.Trim().ToLowerInvariant();
            logs = logs.Where(l =>
                (l.Admin != null && l.Admin.Username.Contains(searchTerm, StringComparison.InvariantCultureIgnoreCase)) ||
                l.Action.Contains(searchTerm, StringComparison.InvariantCultureIgnoreCase) ||
                (l.TargetType != null && l.TargetType.Contains(searchTerm, StringComparison.InvariantCultureIgnoreCase)) ||
                (l.TargetId.HasValue && l.TargetId.Value.ToString().Contains(searchTerm, StringComparison.InvariantCultureIgnoreCase)) ||
                (l.Details != null && l.Details.ToString().Contains(searchTerm, StringComparison.InvariantCultureIgnoreCase))
            ).ToList();
        }

        var total = logs.Count;
        var items = logs
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .ToList();

        return (items, total);
    }
}

File: slp/backend-nestjs/src/modules/admin/admin-log.repository.ts
import { Injectable } from "@nestjs/common";
import { InjectRepository } from "@nestjs/typeorm";
import { Repository } from "typeorm";
import { AdminLog } from "./admin-log.entity";
import { AdminLogFilterDto } from "./dto/admin-log-filter.dto";

export interface CreateLogDto {
  adminId: number;
  action: string;
  targetType?: string;
  targetId?: number;
  details?: any;
}

@Injectable()
export class AdminLogRepository {
  constructor(
    @InjectRepository(AdminLog)
    private readonly repo: Repository<AdminLog>,
  ) {}

  async log(data: CreateLogDto): Promise<void> {
    const log = this.repo.create(data); // TypeORM can handle the object directly
    await this.repo.save(log);
  }

  async getRecent(limit: number = 100): Promise<AdminLog[]> {
    return this.repo.find({
      relations: ["admin"],
      order: { createdAt: "DESC" },
      take: limit,
    });
  }

  async getRecentWithFilterPaginated(
    filter: AdminLogFilterDto,
    page = 1,
    pageSize = 20,
  ): Promise<{ items: AdminLog[]; total: number }> {
    const query = this.repo
      .createQueryBuilder("log")
      .leftJoinAndSelect("log.admin", "admin");

    if (filter.from) {
      query.andWhere("log.createdAt >= :from", { from: filter.from });
    }
    if (filter.to) {
      query.andWhere("log.createdAt <= :to", { to: filter.to });
    }
    if (filter.adminId) {
      query.andWhere("log.adminId = :adminId", { adminId: filter.adminId });
    }
    if (filter.action) {
      query.andWhere("log.action = :action", { action: filter.action });
    }
    if (filter.targetType) {
      query.andWhere("log.targetType = :targetType", {
        targetType: filter.targetType,
      });
    }

    if (filter.search) {
      const searchTerm = `%${filter.search.toLowerCase()}%`;
      query.andWhere(
        `(
        LOWER(admin.username) LIKE :search OR
        LOWER(log.action) LIKE :search OR
        LOWER(log.targetType) LIKE :search OR
        CAST(log.targetId AS TEXT) LIKE :search OR
        CAST(log.details AS TEXT) LIKE :search
      )`,
        { search: searchTerm },
      );
    }

    const total = await query.getCount();

    const items = await query
      .orderBy("log.createdAt", "DESC")
      .skip((page - 1) * pageSize)
      .take(pageSize)
      .getMany();

    return { items, total };
  }
}
